.github/workflows/gradle.yml

```diff
@@ -17,20 +17,93 @@ on:
 jobs:
   build:
 
-    runs-on: macOS-latest
-
     strategy:
       matrix:
-        java: [ 8, 11, 17 ]
+        include:
+          - java: 8
+            platform: macos-latest
+            e2e-tests: android
+          - java: 11
+            platform: macos-latest
+            e2e-tests: ios
+          - java: 17
+            platform: ubuntu-latest
+      fail-fast: false
+
+    runs-on: ${{ matrix.platform }}
 
-    name: JDK ${{ matrix.java }}
+    name: JDK ${{ matrix.java }} - ${{ matrix.platform }}
     steps:
     - uses: actions/checkout@v3
+
     - name: Set up JDK ${{ matrix.java }}
       uses: actions/setup-java@v3
       with:
         distribution: 'zulu'
         java-version: ${{ matrix.java }}
         cache: 'gradle'
+
     - name: Build with Gradle
       run: ./gradlew clean build
+
+    - name: Install Node.js
+      if: ${{ matrix.e2e-tests == 'android' ||  matrix.e2e-tests == 'ios' }}
+      uses: actions/setup-node@v3
+      with:
+        node-version: '18'
+
+    - name: Install Appium
+      if: matrix.e2e-tests == 'android' ||  matrix.e2e-tests == 'ios'
+      run: npm install -g appium@next
+
+    - name: Install UIA2 driver
+      if: matrix.e2e-tests == 'android'
+      run: appium driver install uiautomator2
+
+    - name: Prepare Android emulator
+      if: matrix.e2e-tests == 'android'
+      env:
+        ANDROID_EMU_NAME: test
+        ANDROID_EMU_ABI: x86
+        ANDROID_EMU_TARGET: android-28
+        ANDROID_EMU_TAG: default
+      run: |
+          echo "y" | $ANDROID_HOME/tools/bin/sdkmanager --install "system-images;$ANDROID_EMU_TARGET;$ANDROID_EMU_TAG;$ANDROID_EMU_ABI"
+          echo "no" | $ANDROID_HOME/tools/bin/avdmanager create avd -n "$ANDROID_EMU_NAME" -k "system-images;$ANDROID_EMU_TARGET;$ANDROID_EMU_TAG;$ANDROID_EMU_ABI" --force
+          echo $ANDROID_HOME/emulator/emulator -list-avds
+
+          echo "Starting emulator"
+          nohup $ANDROID_HOME/emulator/emulator -avd "$ANDROID_EMU_NAME" -no-snapshot -delay-adb > /dev/null 2>&1 &
+          $ANDROID_HOME/platform-tools/adb wait-for-device
+          $ANDROID_HOME/platform-tools/adb devices -l
+          echo "Emulator started"
+
+    - name: Run Android E2E tests
+      if: matrix.e2e-tests == 'android'
+      run: ./gradlew uiAutomationTest
+
+    - name: Install XCUITest driver
+      if: matrix.e2e-tests == 'ios'
+      run: appium driver install xcuitest
+
+    - name: Prebuild XCUITest driver
+      if: matrix.e2e-tests == 'ios'
+      run: appium driver run xcuitest build-wda
+
+    - name: Prepare iOS simulator
```

Comment:
@SrinivasanTarget this could also slow down both steps as they are really CPU and I/O intensive and GH VMs are pretty limited in resources. Although the only way to confirm it would be to actually try